src/mux_client_impl/child.rs

```diff
@@ -0,0 +1,112 @@
+use super::{ChildStderr, ChildStdin, ChildStdout, Error};
+
+use core::mem::replace;
+
+use std::io;
+use std::os::unix::process::ExitStatusExt;
+use std::process::ExitStatus;
+
+use openssh_mux_client::connection::{EstablishedSession, SessionStatus};
+
+#[derive(Debug)]
+pub struct RemoteChild {
+    state: RemoteChildState,
+    child_stdin: Option<ChildStdin>,
+    child_stdout: Option<ChildStdout>,
+    child_stderr: Option<ChildStderr>,
+}
+
+impl RemoteChild {
+    pub(crate) fn new(
+        established_session: EstablishedSession,
+        child_stdin: Option<ChildStdin>,
+        child_stdout: Option<ChildStdout>,
+        child_stderr: Option<ChildStderr>,
+    ) -> Self {
+        Self {
+            state: RemoteChildState::Running(established_session),
+            child_stdin,
+            child_stdout,
+            child_stderr,
+        }
+    }
+
+    pub async fn disconnect(self) -> io::Result<()> {
+        use RemoteChildState::*;
+
+        match self.state {
+            Intermediate => unreachable!(),
+            Exited(_exit_status) => Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                "invalid argument: can't kill an exited process",
+            )),
+            Running(_established_session) => Ok(()),
+        }
+    }
+
+    pub async fn wait(&mut self) -> Result<ExitStatus, Error> {
+        use RemoteChildState::*;
+
+        let exit_value = match self.state.take() {
+            Intermediate => unreachable!(),
+            Exited(exit_value) => exit_value,
+            Running(established_session) => match established_session.wait().await {
+                Ok(session_status) => match session_status {
+                    SessionStatus::TtyAllocFail(_established_session) => unreachable!(
+                        "openssh::mux_client_impl does not use feature tty by any means"
+                    ),
+                    SessionStatus::Exited { exit_value } => exit_value,
+                },
+                Err((err, established_session)) => {
+                    self.state = Running(established_session);
+                    return Err(err.into());
+                }
+            },
+        };
+
+        self.state = Exited(exit_value);
+
+        if let Some(val) = exit_value {
+            if val == 127 {
+                Err(Error::Remote(io::Error::new(
+                    io::ErrorKind::NotFound,
+                    "remote command not found",
+                )))
+            } else {
+                Ok(ExitStatusExt::from_raw((val as i32) << 8))
```

Comment:
I forgot to put comment here.

The exit value returned by `openssh_mux_client::EstablishedSession` is the return code of the process, not the exit status returned by `waitpid`.

[The official doc](https://github.com/openssh/openssh-portable/blob/e665ed2d0c24fe11d5470ce72fa1e187377d3fc4/PROTOCOL.mux#L108) also confirms this.

So it is necessary to perform some adjustion here so that it looks like an exit status.

Reply:
Hmm, my reading of that paragraph is actually the opposite of yours. I read that as "the client should exit with the same value as ssh would", and `ssh` returns with the same exit value as the remote process (no adjustment needed).

Reply:
That’s exactly why it needs to be adjusted.

What ssh return is a return value in main, it is different from the exit status retrieved using waitpid, since waitpid would shift the return value in main and ored the killing signal if the process is killed by signal.

Reply:
I don't think the goal here should be to match what waitpid returns. It should be to return the actual exit status of the remote process as best as we can.

Reply:
Yeah, but construction  of ExitStatusExt for unix requires an exit status, not return value of the main function…

Reply:
Also, if I removed the bit shifting there, then the integration test would fail.